### PR TITLE
Added Fitbits

### PR DESCRIPTION
**Pull request by M7-Soliman**

### CODE REVIEW

File: tests/devices/fitbit/test_fitbit_charge_4.py
Comment (TrafficCop) [inline tests/devices/fitbit/test_fitbit_charge_4.py:1]: Good to see the tests are working now! However, I'd suggest also adding more comprehensive tests that, rather than just verifying the data is nonempty, also check the "shape" of the data - i.e. things like whether the start dates of the generated data lie within the requested range, whether the durations make sense (in general making sure generated data points have values that are reasonable).

File: tests/devices/fitbit/test_fitbit_sense.py
Comment (TrafficCop) [inline tests/devices/fitbit/test_fitbit_sense.py:1]: Similar suggestionas the fitbit_charge_4 tests - do some more checks on the data

File: tests/test_all_devices_syn.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def test_all_devices():
         # that they're called the correct number of times.
         # This is essentially checking that the child class has re-implemented
         # the methods correctly.
+        print(device_name)
```
Comment (TrafficCop) [inline tests/test_all_devices_syn.py:37]: This should be removed

File: venv/pyvenv.cfg
Comment (TrafficCop) [inline venv/pyvenv.cfg:1]: Is this file necessary? It looks like it uses a hard coded file directory, which is probably not a good idea for general usage.

File: wearipedia/devices/fitbit/fitbit_sense.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        synthetic_end_date="2022-06-17",
+    ):
+
+        params = {
+            "seed": seed,
+            "synthetic_start_date": synthetic_start_date,
+            "synthetic_end_date": synthetic_end_date,
+        }
+
+        # self.data_types_methods_map = {
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense.py:50]: This can be removed

File: wearipedia/devices/fitbit/fitbit_sense_fetch.py
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_fetch.py:1]: It appears this file is the same as fitbit_charge_4_fetch. Could you make a file such as "fitbit_gen" and then use the same code for both devices? This way we can avoid file bloat.

File: wearipedia/devices/fitbit/fitbit_sense_gen.py
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:1]: Same idea as with the fitbit_sense_fetch. If you can combine this with fitbit_charge_4_gen as one generator file to use between all fitbit devices then that would be preferable.

**Review by TrafficCop — CHANGES_REQUESTED** [review]

File: tests/devices/fitbit/test_fitbit_charge_4.py
Comment (TrafficCop) [inline tests/devices/fitbit/test_fitbit_charge_4.py:1]: Make sure to add a formatted string message for the test cases so if an error does occur, the console logs out the error (i.e. "value should be less than X but was Y")

File: tests/devices/fitbit/test_fitbit_sense.py
Comment (TrafficCop) [inline tests/devices/fitbit/test_fitbit_sense.py:1]: Same comments as in test_fitbit_charge_4.py - add a assert message for the test cases.

File: wearipedia/devices/fitbit/fitbit_sense.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    * `hrv`: heart rate variability data
+    * `distance_day`: distance moved per day detailed by each minute
+
+    :param seed: random seed for synthetic data generation, defaults to 0
+    :type seed: int, optional
+    :param synthetic_start_date: start date for synthetic data generation, defaults to "2022-03-01"
+    :type synthetic_start_date: str, optional
+    :param synthetic_end_date: end date for synthetic data generation, defaults to "2022-06-17"
+    :type synthetic_end_date: str, optional
+    :param single_day: end date for real data data fetching, defaults to "2022-09-19"
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense.py:33]: It looks like this class does not accept single_day as a parameter. If so, you can remove this part of the docstring

File: wearipedia/devices/fitbit/fitbit_charge_4.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        self.minutesVeryActive = syn_data["minutesVeryActive"]
+        self.minutesFairlyActive = syn_data["minutesFairlyActive"]
+        self.minutesLightlyActive = syn_data["minutesLightlyActive"]
+        self.distance = syn_data["distance"]
+        self.minutesSedentary = syn_data["minutesSedentary"]
+
+    def _authenticate(self):
+        # authenticate this device against API
+        fitbit_application()
+        client_id = input("enter client id: ")
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_charge_4.py:113]: Instead of prompting user for client id here, have them send it as a parameter

File: wearipedia/devices/fitbit/fitbit_sense.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        self.distance = syn_data["distance"]
+        self.minutesSedentary = syn_data["minutesSedentary"]
+        self.heart_rate_day = syn_data["heart_rate_day"]
+        self.hrv = syn_data["hrv"]
+        self.distance_day = syn_data["distance_day"]
+
+    def _authenticate(self):
+        # authenticate this device against API
+        fitbit_application()
+        client_id = input("enter client id: ")
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense.py:123]: Instead of prompting user for client id here, have them send it as a parameter

File: wearipedia/devices/fitbit/fitbit_sense_fetch.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+def call_API(access_token: str, url: str, call: str = "GET"):
+    headers = {"Authorization": "Bearer " + access_token}
+    return requests.request(call, url=url, headers=headers).json()
+
+
+def fetch_real_data(
+    data_type, access_token, start_date="", end_date="", single_date=""
+):
+    """Main function for fetching real data from the Fitbit API.
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_fetch.py:14]: add an empty line after this for style purposes

File: wearipedia/devices/fitbit/fitbit_sense_fetch.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+__all__ = ["fetch_real_data"]
+
+
+def call_API(access_token: str, url: str, call: str = "GET"):
+    headers = {"Authorization": "Bearer " + access_token}
+    return requests.request(call, url=url, headers=headers).json()
+
+
+def fetch_real_data(
+    data_type, access_token, start_date="", end_date="", single_date=""
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_fetch.py:12]: start_date and end_date should not have a default value, since these seem to be necessary parameters (lemme know if otherwise).

File: wearipedia/devices/fitbit/fitbit_sense_fetch.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    :param access_token: access token for the API
+    :type api: str
+    :return: the data fetched from the API according to the inputs
+    :rtype: List
+    """
+    single_date = start_date
+
+    categories = {
+        "sleep": {
+            "url": "https://api.fitbit.com/1.2/user/-/sleep/date/"
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_fetch.py:32]: use fstrings for the url of all of these categories

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+    start_phases += timedelta(seconds=arr_of_durations[-1])
+    sleep_dict["levels"]["data"].append(
+        {
+            "dateTime": str(start_phases.isoformat()),
+            "level": "wake",
+            "seconds": round(duration / 60 * percents[2] / 100) * 60,
+        }
+    )
+
+    # do the same for wakeups
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:111]: Same as the other file, I think you should be able to combine this and the previous loop

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+                    "datasetType": "minute",
+                },
+            }
+        ]
+    }
+
+    the_time = time(hour=0, minute=0, second=0)
+    for i in range(1440):
+
+        max_heart_rate = [110, 136, 169, 220]
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:231]: since max_heart_rate, min_dict, weights, are all constants, you should put them outside of the loop as they do not need to be updated

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+                    "dataset": [],
+                    "datasetInterval": 1,
+                    "datasetType": "minute",
+                },
+            }
+        ]
+    }
+
+    the_time = time(hour=0, minute=0, second=0)
+    for i in range(1440):
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:229]: 1440 seems like a magic number, you should rename this to a variable so it is easy to tell what we are looping over. Also, as this appears to be the duration of the exercise, I think you should make it variable rather than fixed (set a random duration).

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+                    "datasetInterval": 1,
+                    "datasetType": "minute",
+                },
+            }
+        ]
+    }
+
+    the_time = time(hour=0, minute=0, second=0)
+
+    for i in range(1440):
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:305]: 1440 as a magic number. Use a variable for duration instead, and name it so it is more readable.

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+    minutesVeryActive = {"dateTime": date, "value": very_active}
+    minutesFairlyActive = {"dateTime": date, "value": fairly_active}
+    minutesLightlyActive = {"dateTime": date, "value": lightly_active}
+    distance = {
+        "dateTime": date,
+        "value": (very_active * 50 + fairly_active * 25 + lightly_active * 10) / 1326,
+    }
+    minutesSedentary = {
+        "dateTime": date,
+        "value": 1440 - (very_active + fairly_active + lightly_active),
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:157]: 1440 is a bit of a mystery number here

Comment (TrafficCop) [reply wearipedia/devices/fitbit/fitbit_sense_gen.py:157]: Name it, and if it is a duration, you should make it variable

File: wearipedia/devices/fitbit/fitbit_sense_gen.py [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+        heart_rate_data["heart_rate_day"][0]["activities-heart"][0]["value"][
+            "heartRateZones"
+        ][inx]["minutes"] += 1
+        heart_rate_data["heart_rate_day"][0]["activities-heart"][0]["value"][
+            "heartRateZones"
+        ][inx]["caloriesOut"] += 2 * (inx + 0.5)
+
+        val = np.random.randint(min_dict[max_heart[0]], max_heart)
+
+        heart_rate_data["heart_rate_day"][0]["activities-heart-intraday"][
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense_gen.py:246]: It looks like here you are generating high frequency heart rate data randomly. I think a better way would be to use a slightly more complex generator so we do not end up with gaussian noise data. You can check out some examples of existing generators at Garmin Fenix or Polar Verity Sense!

File: wearipedia/devices/oura/oura_ring3.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    * `readiness`: readiness data
+    * `heart_rate`: heart_rate
+
+    :param seed: random seed for synthetic data generation, defaults to 0
+    :type seed: int, optional
+    :param synthetic_start_date: start date for synthetic data generation, defaults to "2022-03-01"
+    :type synthetic_start_date: str, optional
+    :param synthetic_end_date: end date for synthetic data generation, defaults to "2022-06-17"
+    :type synthetic_end_date: str, optional
+    :param single_day: end date for real data data fetching, defaults to "2022-09-19"
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3.py:29]: It doesn't look like the single_day parameter is set or used here. You can either remove it from the docstring here or include it in the device

File: wearipedia/devices/fitbit/fitbit_charge_4_gen.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+import numpy as np
+
+__all__ = ["create_syn_data"]
+
+
+def get_sleep(date):
+    """Generate sleep data for a given date.
+
+    :return: sleep data dictionary
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_charge_4_gen.py:13]: include :param for the param it accepts

File: wearipedia/devices/coros/coros_pace_2_gen.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+from random import choice, randrange
+
+import numpy as np
+
+__all__ = ["create_syn_data"]
+
+
+def get_steps(date):
+    """Generate steps data for a given date.
+
```
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:13]: include :param for the param it accepts

File: wearipedia/devices/oura/oura_ring3_authenticate.py [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+            json={
+                "grant_type": "password",
+                "issueRefresh": False,
+                "password": password,
+                "username": email,
+            },
+        )
+
+
+def Oura_token():
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_authenticate.py:26]: lower case Oura as snake case convetion

File: wearipedia/devices/oura/oura_ring3.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+        self.sleep = syn_data["sleep"]
+        self.daily_activity = syn_data["daily_activity"]
+        self.activity = syn_data["activity"]
+        self.ideal_bedtime = syn_data["ideal_bedtime"]
+        self.readiness = syn_data["readiness"]
+        self.heart_rate = syn_data["heart_rate"]
+
+    def _authenticate(self):
+
+        self.user = Oura_token()
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3.py:109]: It looks like you get the user token, but you do't use the login function. is this intended?

File: wearipedia/devices/oura/oura_ring3_authenticate.py [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    """gives us access token given the login info
+    Returns access token"""
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_authenticate.py:28]: you should use a docstring formatted like in the other functions

File: wearipedia/devices/oura/oura_ring3_authenticate.py [hunk 21]
```diff
@@ -0,0 +1,37 @@
+import base64
+import hashlib
+import os
+
+import requests
+
+__all__ = ["login", "Oura_token"]
+
+
+def login(email, password):
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_authenticate.py:10]: This function does not seem to be called anywhere. Is this intended?

File: wearipedia/devices/oura/oura_ring3.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        date_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+        datetime_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+
+        # get the indices by subtracting against the start of the synthetic data
+        synthetic_start = date_str_to_obj(self.init_params["synthetic_start_date"])
+
+        start_idx = (datetime_str_to_obj(params["start_date"]) - synthetic_start).days
+        end_idx = (datetime_str_to_obj(params["end_date"]) - synthetic_start).days
+
+        return data
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3.py:82]: You should filter the data here for the range of provided dates.

File: wearipedia/devices/fitbit/fitbit_sense.py [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+        date_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+        datetime_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+
+        # get the indices by subtracting against the start of the synthetic data
+        synthetic_start = date_str_to_obj(self.init_params["synthetic_start_date"])
+
+        start_idx = (datetime_str_to_obj(params["start_date"]) - synthetic_start).days
+        end_idx = (datetime_str_to_obj(params["end_date"]) - synthetic_start).days
+
+        return data
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_sense.py:90]: You should filter the data here for the range of provided dates.

File: wearipedia/devices/fitbit/fitbit_charge_4.py [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+        date_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+        datetime_str_to_obj = lambda x: datetime.strptime(x, "%Y-%m-%d")
+
+        # get the indices by subtracting against the start of the synthetic data
+        synthetic_start = date_str_to_obj(self.init_params["synthetic_start_date"])
+
+        start_idx = (datetime_str_to_obj(params["start_date"]) - synthetic_start).days
+        end_idx = (datetime_str_to_obj(params["end_date"]) - synthetic_start).days
+
+        return data
```
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_charge_4.py:84]: You should filter the data here for the range of provided dates.

File: wearipedia/devices/oura/oura_ring3_fetch.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
+import requests
+
+__all__ = ["fetch_real_data"]
+
+
+def fetch_real_data(data_type, access_token, start_date, end_date):
+    """Main function for fetching real data from the Fitbit API.
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_fetch.py:7]: add an empty line to space this line

File: wearipedia/devices/oura/oura_ring3_fetch.py [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+            start_date_col: start_date,
+            end_date_col: end_date,
+        }
+        return requests.request(call, url=url, params=params).json()
+
+    def call_API_version_2(
+        url: str,
+        start_date=start_date,
+        end_date=end_date,
+        access_token=access_token,
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_fetch.py:40]: For consistency, you should list the data type of the parameter for all parameters or don't have them

File: wearipedia/devices/oura/oura_ring3_fetch.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+        call: str = "GET",
+    ):
+        params = {
+            "access_token": access_token,
+            start_date_col: start_date,
+            end_date_col: end_date,
+        }
+        return requests.request(call, url=url, params=params).json()
+
+    def call_API_version_2(
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_fetch.py:36]: I would rename this function to follow snake case convention (now capital API)

File: wearipedia/devices/oura/oura_ring3_fetch.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+    :type end_date: str
+    :param data_type: the type of data to fetch, one of "Personal Info", "Heart Rate", "Sessions", "Tags", "Workouts", "Daily Sleep", "Daily Activity", "Daily Readiness", "Ideal Bedtime"
+    :type data_type: str
+    :param access_token: access token for the API
+    :type api: str
+    :return: the data fetched from the API according to the inputs
+    :rtype: List
+    """
+
+    def call_API_version_1(
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_fetch.py:20]: I would rename this function to follow snake case convention (now capital API)

File: wearipedia/devices/oura/oura_ring3_fetch.py [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+        end_date=end_date,
+        access_token=access_token,
+        start_date_col: str = "start_date",
+        end_date_col: str = "end_date",
+        call: str = "GET",
+    ):
+        headers = {"Authorization": "Bearer " + access_token}
+        params = {start_date_col: start_date, end_date_col: end_date}
+        return requests.request(call, url=url, headers=headers, params=params).json()
+
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_fetch.py:48]: I think the two call api functions are a little confusing. I think you should include a docstring for each function to explain what they do differently or add an explanatory comment above both.

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 30]
```diff
@@ -0,0 +1,344 @@
...
+from datetime import datetime, timedelta
+
+import numpy as np
+
+__all__ = ["create_syn_data"]
+
+
+def get_daily_activity(date):
+    """Generate daily activity data for a given date.
+
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:12]: include a parameters docstring

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 31]
```diff
@@ -0,0 +1,344 @@
...
+
+def get_heart_rate(date):
+    date = datetime.strptime(date, "%Y-%m-%d")
+
+    timestamp = date.timestamp()
+
+    heart_rate_data = []
+
+    for i in range(0, 24 * 60, 5):
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:279]: The range of your loop contains a magic number. You should make this variable and named instead

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 32]
```diff
@@ -0,0 +1,344 @@
...
+        bpm = random.randint(50, 120)
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:280]: use a generator for this as opposed to random noise - i.e. Garmin Fenix 7s, Polar Verity Sense

File: wearipedia/devices/oura/oura_ring3_gen.py
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:1]: Please include :param field for all docstrings to explain the parameter it accepts

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 33]
```diff
@@ -0,0 +1,344 @@
...
+
+data = create_syn_data("2022-05-02", "2022-05-5")
+
+heart_rate = data["heart_rate"]
+
+bpm = []
+for elem in heart_rate:
+    bpm.append(elem["bpm"])
+
+print(bpm)
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:344]: It does not look like this print statement is necessary

File: wearipedia/devices/coros/coros_pace_2_gen.py [hunk 34]
```diff
@@ -0,0 +1,307 @@
...
+    return data
+
+
+def get_sports(date_str):
+    data = {"sports": {"apiCode": "B7D57DB7", "data": []}}
+
+    date = datetime.strptime(date_str, "%Y-%m-%d")
+
+    for _ in range(random.randint(1, 5)):
```
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:141]: Not sure what 1,5 represents. Use a variable name to indicate it

File: wearipedia/devices/coros/coros_pace_2_gen.py
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:1]: I see in your coros_pace_2 notebook you have a visualization for daily calories, but there is no generator for that here. Could you add it?

**Review by TrafficCop — CHANGES_REQUESTED** [review]
Please check off completed reviews as you go

File: wearipedia/devices/fitbit/fitbit_sense_gen.py (hunk 9 above)
Comment (TrafficCop) [reply wearipedia/devices/fitbit/fitbit_sense_gen.py:112]: Please address this

**Review by TrafficCop — CHANGES_REQUESTED** [review]
Please address all comments and resolve them once complete

File: wearipedia/devices/fitbit/fitbit_sense_fetch.py [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+
+def fetch_real_data(data_type, access_token, start_date, end_date, single_date):
+    """Main function for fetching real data from the Fitbit API.
+
+    :param start_date: the start date represented as a string in the format "YYYY-MM-DD"
+    :type start_date: str
+    :param end_date: the end date represented as a string in the format "YYYY-MM-DD"
+    :type end_date: str
+    :param single_date: the day data is being requested from represented as a string in the format "YYYY-MM-DD"
+    :type end_date: str
```
[the remark this answers is not in the document]
Comment (M7-Soliman) [reply wearipedia/devices/fitbit/fitbit_sense_fetch.py:19]: no not really, I just have to remove te single date

File: wearipedia/devices/oura/oura_ring3_fetch.py (hunk 26 above)
Comment (M7-Soliman) [reply wearipedia/devices/oura/oura_ring3_fetch.py:40]: all already has a datatype

File: wearipedia/devices/coros/coros_pace_2_gen.py
Comment (M7-Soliman) [reply wearipedia/devices/coros/coros_pace_2_gen.py:1]: Its the calorie line

File: wearipedia/devices/coros/coros_pace_2.py [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+        self.exercise_time = syn_data["exercise_time"]
+        self.heart_rate = syn_data["heart_rate"]
+        self.sports = syn_data["sports"]
+        self.sleep = syn_data["sleep"]
+        self.active_energy = syn_data["active_energy"]
+
+    def _authenticate(self):
+        # authenticate this device against API
+        print(
+            "Get the access token here: https://colab.research.google.com/drive/1cnVktqcPXPtGS6JcdHVVCqIC3nngfxhm?usp=sharing"
```
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2.py:112]: This takes the user to the coros notebook, where can they get the access token?

File: wearipedia/devices/coros/coros_pace_2_gen.py [hunk 37]
```diff
@@ -0,0 +1,354 @@
...
+
+def get_steps(date):
+    """Returns a an array of steps data
+
+    :param date: the date as a string in the format "YYYY-MM-DD"
+    :type start_date: str
+    :rtype: dictionary
+    """
+
+    """Generate steps data for a given date.
```
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:21]: You only need one doc string per function, you can remove the other

File: wearipedia/devices/coros/coros_pace_2_gen.py [hunk 38]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    return data
+
+
+def get_sports(date_str):
+    """Returns a an array of sports data
+
+    :param date: the date as a string in the format "YYYY-MM-DD"
+    :type start_date: str
```
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:165]: get_sports does not have a start_date field

File: wearipedia/devices/coros/coros_pace_2_gen.py
Comment (TrafficCop) [inline wearipedia/devices/coros/coros_pace_2_gen.py:1]: Please review all of your docstrings **carefully**. It appears that they were copy and pasted and have different variables to what are accepted.

File: wearipedia/devices/fitbit/fitbit_charge4_fetch.py
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_charge4_fetch.py:1]: Please address the changes I listed from the comments before resolving

File: wearipedia/devices/fitbit/fitbit_charge_4_gen.py
Comment (TrafficCop) [inline wearipedia/devices/fitbit/fitbit_charge_4_gen.py:1]: For all docstrings here, rather than just saying "rtype: dictionary" or "rtype: defaultdict," be more descriptive about the format and do "rtype: Dict[list[int]]" (what the data structure is, what it contains within it

File: wearipedia/devices/oura/oura_ring3_gen.py
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:1]: Please **carefully** review all the docstrings here. Some of them list a "start_date" type even though the function doesn't accept a "start_date." Also, for your return types, please be descriptive about what is contained within each data structure i.e. "Dictionary[list[int]]" instead of "dictionary" for a dictionary containing lists of ints

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 39]
```diff
@@ -0,0 +1,376 @@
...
+        full_dict["daily_activity"].append(get_daily_activity(date))
+        full_dict["activity"].append(get_activity(date))
+        full_dict["ideal_bedtime"].append(get_ideal_bedtime(date))
+        full_dict["readiness"].append(get_readiness(date))
+        full_dict["heart_rate"].extend(get_heart_rate(date))
+
+    return full_dict
+
+
+print(get_heart_rate("2011-12-06"))
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:376]: you can remove this

File: wearipedia/devices/oura/oura_ring3_gen.py [hunk 40]
```diff
@@ -0,0 +1,376 @@
...
+    ideal_bedtime_data = {
+        "date": date.strftime("%Y-%m-%d"),
+        "bedtime_window": {"start": start_time, "end": end_time},
+        "status": "IDEAL_BEDTIME_AVAILABLE",
+    }
+
+    return ideal_bedtime_data
+
+
+def get_heart_rate(date):
```
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_gen.py:292]: missing docstring here

File: wearipedia/devices/oura/oura_ring3_authenticate.py
Comment (TrafficCop) [inline wearipedia/devices/oura/oura_ring3_authenticate.py:1]: Is it necessary to have this function in an extra file? It might be a good idea to move it into the authenticate function for oura_ring3 since it is not called anywhere else

**Review by TrafficCop — CHANGES_REQUESTED** [review]
Mostly just style changes, please address them carefully before resolving.

**Review by TrafficCop — APPROVED** [review]
